### PR TITLE
feat: new hooks, useConditionalEffect and useConditionalUpdateEffect

### PR DESCRIPTION
**Pull request by xobotyi**
## What new hook does?

- `useConditionalEffect`  - Alike `useEffect` but callback invoked only if conditions match predicate.
- `useConditionalUpdateEffect` — Alike useUpdateEffect but callback invoked only if conditions match predicate.

## Checklist

- [x] Have you read [contribution guideline](../../CONTRIBUTING.md)?
- [x] Does the code have comments in hard-to-understand areas?
- [ ] Is there an existing issue for this PR?
  - _link issue here_
- [x] Have the files been linted and formatted?
- [x] Have the docs been updated?
- [x] Have the tests been added to cover new hook?
- [x] Have you run the tests locally to confirm they pass?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,6 +60,10 @@ import { useMountEffect } from "@react-hookz/web/esnext";
     — Run effect only when component is unmounted.
   - [`useUpdateEffect`](https://react-hookz.github.io/web/?path=/docs/lifecycle-useupdateeffect)
     — Effect hook that ignores the first render (not invoked on mount).
+  - [`useConditionalEffect`](https://react-hookz.github.io/web/?path=/docs/lifecycle-useconditionaleffect)
+    — Alike `useEffect` but callback invoked only if conditions match predicate.
```
Comment (JoeDuncko) [inline README.md:64]: "Like" probably reads better than "Alike" here.

Comment (JoeDuncko) [reply README.md:64]: Same with everywhere else.

File: src/useConditionalUpdateEffect.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  const shouldInvoke = !isFirstMount && predicate(conditions);
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  const deps = useRef<{}>();
+
+  // we want callback invocation only in case all conditions matches predicate
+  if (shouldInvoke) {
+    deps.current = {};
+  }
+
+  // we cant avoid on-mount invocations so slip noop callback for the cases we dont need invocation
```
Comment (JoeDuncko) [inline src/useConditionalUpdateEffect.ts:35]: can't

File: stories/useConditionalUpdateEffect.story.mdx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import { Example } from "./useConditionalUpdateEffect.stories";
+
+<Meta title="Lifecycle/useConditionalUpdateEffect" component={Example} />
+
+# useConditionalUpdateEffect
+
+Alike `useUpdateEffect` but callback invoked only if conditions match predicate. By default, predicate
+matches if all conditions are truthy.
+
+As `useUpdateEffect`, it _does not_ perform conditions match on first mount.
```
Comment (JoeDuncko) [inline stories/useConditionalUpdateEffect.story.mdx:11]: "As" -> "Like", I think

File: stories/useConditionalUpdateEffect.story.mdx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  callback: React.EffectCallback,
+  conditions: T,
+  predicate?: IUseConditionalUpdateEffectPredicate<T>
+): void;
+```
+
+#### Arguments
+
+- _**callback**_ _`React.EffectCallback`_ - Effect callback like for `React.useEffect` hook
+- _**conditions**_ _`ReadonlyArray<any>`_ - Conditions that matched against predicate
```
Comment (JoeDuncko) [inline stories/useConditionalUpdateEffect.story.mdx:36]: "Conditions that are matched against predicate", maybe

**Review by JoeDuncko — CHANGES_REQUESTED** [review]

File: stories/useConditionalUpdateEffect.story.mdx (hunk 4 above)
Comment (xobotyi) [reply stories/useConditionalUpdateEffect.story.mdx:36]: yup

**Comment by JoeDuncko** [comment]
Every time you go to the storybook example at http://localhost:6006/?path=/docs/lifecycle-useconditionaleffect--example it shows this dialogue. It'll probably get annoying. Can we make a different example for that?

image

**Comment by JoeDuncko** [comment]
Maybe like I did for http://localhost:6006/?path=/docs/lifecycle-useunmounteffect--example where we have a button to manually "mount" the component?

**Comment by xobotyi** [comment]
Maybe, i've also got annoyed but had no quick ideas about different exaple

**Comment by JoeDuncko** [comment]
> Maybe, i've also got annoyed but had no quick ideas about different exaple

Should we block on making a different example then? Or merge anyway?

**Comment by xobotyi** [comment]
Already doing that

**Review by JoeDuncko — APPROVED** [review]

**Comment by JoeDuncko** [comment]
Ah, approved before I saw your reply. Will re-review when you have another example up. Or just feel free to merge.

**Comment by JoeDuncko** [comment]
Careful with those force pushes. I'm testing locally, so when you force push it rewrites history and it's hard to reconcile on my machine unless I know to rebase. Unless you have suggestions for handling that?

**Comment by xobotyi** [comment]
I've just rebased to master, it has conflicts.

**Comment by xobotyi** [comment]
There are some side changes within docs, i've done it here to not to waste the time for separate PRs %)

**Comment by JoeDuncko** [comment]
LGTM

**Pull request merged**